[#6] OAuth2 소셜 로그인, 회원가입 및 JWT 인증 구현

khoon9 opened the pull request:
## 목표
- Issue #6 해결
- Google OAuth2 + JWT 기반 소셜 로그인·회원가입 기능 제공

## 수행 내용
1. 빌드 및 설정
- spring-boot-starter-security, spring-boot-starter-oauth2-client, jjwt 의존성 추가
- application.yml에 OAuth Client·JWT secret·TTL 등 환경 변수화
2. AuthAccount entity 추가
- OAuth2 인증 정보 관리용 entity
3. User entity 필드 추가
- 사용자 UUID 기반 외부용 식별자, 이메일 필드 추가
4. 인증 및 인가 로직 추가
- OAuth2 프로필 수신 -> 신규 사용자 자동 저장
- 로그인 성공 시 JWT 발급 후 리다이렉트
- 요청 헤더의 JWT 검증 -> 인증 컨텍스트 주입
5. 보안 설정
- 인증 예외 범위 설정: /, /oauth2/**, /login/** → permitAll
- 인증 범위 설정: 그 외 모든 요청(/api/**) ->authenticated
- 세션 정책 STATELESS, 커스텀 JWT 필터 체인 등록

## 유의
- CSRF : 쿠키 기반 세션이 없고 Bearer JWT만 사용하므로 csrf().disable() 적용

## 시퀀스 다이어그램
- OAuth2 로그인 및 회원가입 & JWT 발급
<img width="1701" alt="스크린샷 2025-05-31 오후 10 53 44" src="https://github.com/user-attachments/assets/bf946f21-4a4a-4e0d-b2b6-709061674a49" />

-  보호 API 호출 시 JWT 인증
<img width="1266" alt="스크린샷 2025-05-31 오후 11 04 37" src="https://github.com/user-attachments/assets/c683be02-97a0-4614-9e51-8c8874e42168" />

khoon9 committed:
build: Security-OAuth2, JWT 의존성 추가

khoon9 committed:
chore: application.yml 업데이트

- oauth2-client 설정, 프론트엔드 리다이렉트 URI, JWT 서명키·만료시간 추가

khoon9 committed:
feat: AuthProvider enum 추가

- 구글 이외의 또다른 소셜 로그인 추가 대비

khoon9 committed:
feat: User entity 필드 추가, UserRepository 조회 메서드 추가

- User 엔티티 provider, providerId, email 필드 추가
- UserRepository 클래스에 provider와 providerId 필드 기반 사용자 조회 메서드 추가

khoon9 committed:
feat: JwtTokenProvider 구현

- @PostConstruct 에서 HS256 SecretKey 초기화
- createToken(subject, role) -> JWT 발급
- parseClaims(token) -> 서명 검증 및 Claims 반환

khoon9 committed:
feat: CustomOAuth2UserService 구현

1. provider + providerId 로 사용자 조회
2. 조회된 사용자 없으면 User 생성
3. DefaultOAuth2User 반환(ROLE_USER)

khoon9 committed:
feat: OAuth2SuccessHandler에서 JWT 발급 후 프론트로 리다이렉트

- 로그인 성공 -> User 조회 -> JWT 생성
- 토큰 파라미터 설정 후 프론트로 리다이렉트

khoon9 committed:
feat: CustomUserDetails 정의

- userId, username(subject), 권한 컬렉션 필드
- 패스워드 공백(소셜 로그인)

khoon9 committed:
feat: CustomUserDetailsService 구현

- JWT 인증 필터가 전달한 subject(provider:providerId)로 DB에서 사용자 조회
- Spring Security 가 요구하는 UserDetailsService 를 커스텀해 일관된 Principal 생성

khoon9 committed:
feat: JwtAuthenticationFilter 추가

- Authorization: Bearer 토큰 파싱
- 토큰 검증 실패 시 RuntimeException 발생
- SecurityContext 에 UsernamePasswordAuthenticationToken 저장

khoon9 committed:
feat: SecurityConfig 추가

- CSRF 비활성, 세션 STATELESS
- 퍼블릭 엔드포인트: "/", "/oauth2/**", "/login/**"
- OAuth2 로그인: customUserService + successHandler
- jwtAuthenticationFilter → UsernamePasswordAuthenticationFilter 앞에 배치

khoon9 committed:
chore: 매개변수 이름 변경

- 입력 값이 email 형식이 아니기에, 의미를 반영하기 위해 subject로 수정

khoon9 committed:
feat: JWT 예외 처리 로직 추가

- JWT 예외별 errorCode 메시지를 정의하고
ApiErrorResponse로 응답 형식을 통일함

f-lab-green commented on src/main/java/flab/transtalk/auth/service/CustomOAuth2UserService.java:
```diff
@@ -0,0 +1,56 @@
+                providerId = (String) oAuth2User.getAttribute("sub");
+                nameAttributeKey = "sub";
+            }
+            default -> throw new OAuth2AuthenticationException("Unsupported provider: " + registrationId);
```
사소하긴 하지만 이게 클라이언트까지 날아갈 수 있는 정보라면 공격자가 해당 서비스가 사용하는 프로바이더를 추측할 수 있을거 같아요. (단순 코멘트 입니다.)

khoon9 replied:
넵 에러 메시지를 보다 포괄적인 내용으로 수정하여 반영했습니다

f-lab-green commented on src/main/java/flab/transtalk/auth/service/CustomOAuth2UserService.java:
```diff
@@ -0,0 +1,56 @@
+                        .providerId(providerId)
+                        .email(email)
+                        .name(name)
+                        .build()));
```
https://github.com/f-lab-edu/trans-talk/blob/0e6aabeb624c209ebdfcad9bedb07ea740c9c14e/src/main/java/flab/transtalk/user/domain/User.java#L47

여기 생일이 있었던거 같은데요.

khoon9 replied:
생일을 입력받기 위해선 구글 내에서 별도의 인증 과정이 추가로 필요하다고 하여, null 값이 입력되도록 수정했습니다. 추후 이 부분을 구체화하여 수정할계획입니다!

khoon9 replied:
생일을 입력받기 위해선 구글 내에서 별도의 인증 과정이 추가로 필요하다고 하여, null 값이 입력되도록 수정했습니다. 추후 이 부분을 구체화하여 수정할 계획입니다!

f-lab-green commented on src/main/java/flab/transtalk/auth/service/CustomOAuth2UserService.java:
```diff
@@ -0,0 +1,56 @@
+
+        return new DefaultOAuth2User(
+                Collections.singleton(new SimpleGrantedAuthority(DEFAULT_ROLE)),
+                oAuth2User.getAttributes(),
```
filter-out 해서 필요한 정보만 내보내는건 어떨까요.

f-lab-green commented on src/main/java/flab/transtalk/auth/security/oauth/OAuth2SuccessHandler.java:
```diff
@@ -0,0 +1,62 @@
+        String jwt = jwtTokenProvider.createToken(subject, DEFAULT_ROLE);
+
+        String redirectUrl = UriComponentsBuilder.fromUriString(this.redirectUri)
+                .queryParam("token", jwt)
```
민감 데이터인지 확인해보는게 좋을 듯 합니다.

khoon9 replied:
기존 provider:providerId 형식의 subject를 제거하고, UUID 기반 사용자 식별자(externalId)를 사용하는 방식으로 전환하여 token에 포함된 내용 중 사용자를 특정할 수 있는 정보를 제거하였습니다. 감사합니다.

f-lab-green commented on src/main/java/flab/transtalk/auth/security/oauth/OAuth2SuccessHandler.java:
```diff
@@ -0,0 +1,62 @@
+                .findByProviderAndProviderId(provider, providerId)
+                .orElseThrow();
+
+        String subject = user.getProvider() + ":" + user.getProviderId();
```
user에 getSubject 같은 로직으로 만드는게 좋아보여요. 나중에 규칙이 변경되었을 때 케이스를 모두 찾아야하니까요.

khoon9 replied:
넵 말씀주신 내용 반영해서 인증 정보를 별도로 담당하는 ProviderInfo dto를 만들어서 fromUser, fromSubject, toSubject와 같은 메소드를 정의하는 방식으로 접근하도록 하겠습니다.

khoon9 replied:
넵 말씀주신 내용 반영해서 인증 정보를 별도로 담당하는 ProviderInfo 클래스를 만들어서 fromUser, fromSubject, toSubject와 같은 메소드를 정의하는 방식으로 접근하도록 하겠습니다.

khoon9 replied:
기존 provider:providerId 형식의 subject를 제거했습니다. 이에 따라 ProviderInfo 클래스의 fromUser, fromSubject, toSubject 메소드 또한 코드에서 제외했습니다.

khoon9 replied:
기존 provider:providerId 형식의 subject를 제거하고 별도의 사용자 식별자(UUID 기반 임의 값)를 사용하는 것으로 전환하였습니다. 이에 따라 ProviderInfo 클래스의 fromUser, fromSubject, toSubject 메소드 또한 코드에서 제외하여 반영했습니다.

khoon9 replied:
기존 provider:providerId 형식의 subject를 제거하고 별도의 사용자 식별자(UUID 기반 임의 값)를 사용하는 것으로 전환하였습니다. 이에 따라 ProviderInfo 클래스의 fromUser, fromSubject, toSubject 메소드 또한 코드에서 제거하여 반영했습니다.

f-lab-green commented on src/main/java/flab/transtalk/auth/security/oauth/OAuth2SuccessHandler.java:
```diff
@@ -0,0 +1,62 @@
+
+        User user = userRepository
+                .findByProviderAndProviderId(provider, providerId)
+                .orElseThrow();
```
이 에러를 핸들링하는 곳이 있을까요?

khoon9 replied:
누락된 걸 확인하여, 추가했습니다. 감사합니다

f-lab-green commented on src/main/java/flab/transtalk/auth/security/oauth/OAuth2SuccessHandler.java:
```diff
@@ -0,0 +1,62 @@
+
+        AuthProvider provider;
+        String nameAttributeKey;
+        switch (registrationId) {
```
이 구조가 반복되는 것으로 보이는데 resolver를 만들어보는 것도 좋아보여요.

khoon9 replied:
ProviderInfo 클래스에서 fromOAuth2User 메소드를 통해 OAuth2User 객체로부터 provider와 providerId를 정보를 추출할 수 있도록 구성하였습니다!

f-lab-green commented on src/main/java/flab/transtalk/auth/service/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,51 @@
+    @Override
+    @Transactional(readOnly = true)
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        if (!username.contains(":")) {
```
앞선 코멘트와 동일합니다. 하드코딩된 것이 변경에 취약하므로 파싱을 담당하는 객체를 별도로 두는 것이 좋아보여요.

khoon9 replied:
넵! Provider와 관련된 로직을 ProviderInfo 클래스로 캡슐화하여 반영하였습니다.

f-lab-green commented on src/main/java/flab/transtalk/auth/security/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+        String header = request.getHeader("Authorization");
```
스프링 HttpHeaders에 정의된 것을 사용하는 것이 안전해요.
```java
HttpHeaders.AUTHORIZATION
```

khoon9 replied:
넵 반영했습니다

f-lab-green reviewed (commented):
수고 많으셨습니다. 코멘트 달았어요!
그리고 빌드가 또 깨지는데 빌드는 항상 완성된 것을 보냅시다.

khoon9 committed:
fix: 소나 스크립트에서 불필요한 환경변수 설정 제거

- 코드 정적 분석에 필요없는 설정임을 확인하여 제거하였음

khoon9 commented on src/main/java/flab/transtalk/user/domain/User.java:
```diff
@@ -27,6 +33,16 @@ public class User {
+    private String providerId;
+
+    @Column(name = "email")
+    private String email;
```
> 그리고 시스템 내에서 email이 uq constraint를 가질 필요는 혹시 없는걸까요?

이후 필요에 따라 제공자를 확장하는 과정에서 email이 중복되는 것을 염려하여 uq constraint을 걸지 않은 상태입니다. 시스템 내에서 email의 역할이 아직까진 명확히 확정되지 않은 상태이기에 만약 제약조건이 추가로 필요하다면 그에 따라 바꾸어나갈 방침입니다.

khoon9 committed:
fix: 사용자 매칭 로직 변경

- 매칭 대상이 탐색되 않더라도 정상범주로 처리하도록 방침 변경
- 매칭 대상은 최대 4 명, 추후 변경되는 알고리즘에 의해 구체화 예정
- MySQL에 종속성이 있는 쿼리문 수정

khoon9 commented on src/main/java/flab/transtalk/user/repository/UserRepository.java:
```diff
@@ -1,15 +1,19 @@
+import java.util.Optional;
 
 public interface UserRepository extends JpaRepository<User, Long> {
     @Query("SELECT u FROM User u WHERE u.id <> :userId ORDER BY function('RAND')")
```
넵! 기존 ORDER BY function('RAND')을 제거하고, Pageable을 도입하여 offset을 서비스 레이어에서 임의 설정하는 방식으로 수정하여 반영했습니다.

khoon9 committed:
chore: secret key 인코딩 방식 명시

- UTF-8 encoding 사용

khoon9 committed:
fix: OAuth2 exception message 수정

- provider 유추 가능성으로인해 에러 메시지를 보다 포괄적인 내용으로 수정

khoon9 committed:
refactor: HttpHeaders constant 사용

- auth header를 읽는 과정에서 기존 하드코딩 방식이 아닌 HttpHeaders constant을 사용하여 불러오도록 수정

khoon9 committed:
refactor: OAuth2 authentication process 수정

- Improves OAuth2 authentication flow by introducing a ProviderInfo class to encapsulate provider-related logic.

khoon9 committed:
fix: user not found에 해당하는 예외 명시

- 사용자 발견 실패로 인한, 인증 실패 예외 Throw 명시

khoon9 committed:
refactor: JWT parsing exception 구조 수정

- 에러 발생 가능 로직과 되도록 가까운 곳에서 에러 메시지를 처리하도록 구조 수정

khoon9 committed:
fix: principal 포함 정보 제한하도록 수정

- principal에 포함될 속성값들을 filter-out하여 정말 필요한 정보만 시큐리티 컨텍스트에서 다루도록 수정

khoon9 committed:
refactor: Jwt claim 확장 용이하도록 수정

- 기존 개별 입력 구조에서 Map 자료형 기반 입력으로 구조적 변경 수행

khoon9 commented on src/main/java/flab/transtalk/auth/security/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,45 @@
+        Date now = new Date();
+        return Jwts.builder()
+                .setSubject(email)
+                .claim("role", role)
```
Jwt claim의 확장이 용이하도록 .addClaims 메소드를 통해 Map 자료형을 사용하여 한번에 입력되도록 수정하여 반영하였습니다.

khoon9 committed:
refactor: jwt subject를 externalId 기반으로 전환

- jwt에 포함되던 민감한 내용을 제거하고 서버 내의 고유 식별 Id(임의 생성 UUID)로 대체
- 기존 provider:providerId 형식의 subject를 제거하고, UUID 기반 사용자 식별자(externalId)를 사용하는 방식으로 전환하였습니다.

khoon9 commented on src/main/java/flab/transtalk/auth/security/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,45 @@
+        return Jwts.parserBuilder()
+                .setSigningKey(key)
+                .build()
+                .parseClaimsJws(token);
```
> 1. 토큰 만료나 위조가 있으면 예외를 던지던 친구로 기억해요. 확인해보고 예외처리를 다는 것이 좋을 것 같아요.

try/catch로 묶어서 토큰 만료나 위조 예외 발생시 예외를 throw 하도록 수정하였습니다.


> 2. Jws는 외부 라이브러리의 스펙이기 때문에 애플리케이션의 로직으로 감싸거나 validation 완료 여부만 반환해도 좋아보여요.

넵! Jws<Claims>을 반환하던 parseClaims 메소드를 private로 수정하고 클래스 내에서 자체적으로 호출하여 getSubject을 통해 subject를 반환하도록 구현하였습니다.

khoon9 replied:
> 1. 토큰 만료나 위조가 있으면 예외를 던지던 친구로 기억해요. 확인해보고 예외처리를 다는 것이 좋을 것 같아요.

try/catch로 묶어서 토큰 만료나 위조 예외 발생시 예외를 throw 하도록 수정하였습니다.


> 2. Jws는 외부 라이브러리의 스펙이기 때문에 애플리케이션의 로직으로 감싸거나 validation 완료 여부만 반환해도 좋아보여요.

넵! Jws 타입을 반환하던 parseClaims 메소드를 private로 수정하고 클래스 내에서 자체적으로 호출하여 getSubject을 통해 subject를 반환하도록 구현하였습니다.

khoon9 committed:
refactor: Jws 의존성 관리

- Jws는 외부 라이브러리의 스펙이기 때문에 의존성 관리를 위하여 JwtTokenProvider 내에서 자체적으로 처리하도록 리팩터링 수행

khoon9 committed:
refactor: 사용자 엔티티로부터 인증 정보 분리

- OAuth2 인증 정보를 사용자 엔티티(User)와 분리하여 관리하기 위해 AuthAccount 엔티티를 도입함
- User 엔티티에서 provider 및 providerId 필드를 제거하였고, OAuth2SuccessHandler 및 CustomOAuth2UserService에서 인증 계정 조회 및 생성을 위해 AuthAccountRepository를 사용하도록 수정함

khoon9 commented on src/main/java/flab/transtalk/user/domain/User.java:
```diff
@@ -27,6 +33,16 @@ public class User {
 
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false)
+    private AuthProvider provider;
```
말씀주신 내용 반영하여, 사용자 엔티티로부터 인증 정보를 분리하는 작업을 수행하였습니다! 감사합니다.

khoon9 committed:
refactor: JWT exception handling with error codes

- JWT의 특정 예외들을 JwtErrorCode으로 구조화 하여 관리

khoon9 committed:
fix: 불필요한 중복 인스턴스 생성 코드 제거

- Random 인스턴스를 매번 생성하지 말고 재사용하도록 수정

khoon9 committed:
style: 사용하지 않는 import 구문 제거

khoon9 committed:
style: collect(Collectors.toList()) 대신 toList() 사용

khoon9 committed:
style: 사용하지 않는 변수 및 import 구문 제거
